diff -Naur busybox-1.01.orig/editors/vi.c busybox-1.01/editors/vi.c
--- busybox-1.01.orig/editors/vi.c	2005-08-17 09:29:16.000000000 +0800
+++ busybox-1.01/editors/vi.c	2009-12-31 08:58:48.756427965 +0800
@@ -19,7 +19,7 @@
  */
 
 static const char vi_Version[] =
-	"$Id: vi.c,v 1.38 2004/08/19 19:15:06 andersen Exp $";
+	"$Id: vi.c,v 1.35 2004/03/31 11:12:51 andersen Exp $";
 
 /*
  * To compile for standalone use:
@@ -344,7 +344,6 @@
 #endif							/* CONFIG_FEATURE_VI_CRASHME */
 
 	status_buffer = STATUS_BUFFER;
-	*status_buffer = '\0';	// clear status buffer
 
 #ifdef CONFIG_FEATURE_VI_READONLY
 	vi_readonly = readonly = FALSE;
@@ -463,6 +462,7 @@
 	last_forward_char = last_input_char = '\0';
 	crow = 0;
 	ccol = 0;
+	edit_status();
 
 #ifdef CONFIG_FEATURE_VI_USE_SIGNALS
 	catch_sig(0);
@@ -506,7 +506,6 @@
 	adding2q = 0;
 #endif							/* CONFIG_FEATURE_VI_DOT_CMD */
 	redraw(FALSE);			// dont force every col re-draw
-	edit_status();
 	show_status_line();
 
 	//------This is the main Vi cmd handling loop -----------------------
@@ -872,7 +871,6 @@
 		} else {
 			// user wants file status info
 			edit_status();
-			show_status_line();
 		}
 	} else if (strncasecmp((char *) cmd, "features", i) == 0) {	// what features are available
 		// print out values of all features
@@ -1335,30 +1333,22 @@
 {
 	if (dot > text && dot[-1] != '\n')
 		dot--;
-	edit_status();			// show current file status
-	show_status_line();
 }
 
 static void dot_right(void)
 {
 	if (dot < end - 1 && *dot != '\n')
 		dot++;
-	edit_status();			// show current file status
-	show_status_line();
 }
 
 static void dot_begin(void)
 {
 	dot = begin_line(dot);	// return pointer to first char cur line
-	edit_status();			// show current file status
-	show_status_line();
 }
 
 static void dot_end(void)
 {
 	dot = end_line(dot);	// return pointer to last char cur line
-	edit_status();			// show current file status
-	show_status_line();
 }
 
 static Byte *move_to_col(Byte * p, int l)
@@ -1383,15 +1373,11 @@
 static void dot_next(void)
 {
 	dot = next_line(dot);
-	edit_status();			// show current file status
-	show_status_line();
 }
 
 static void dot_prev(void)
 {
 	dot = prev_line(dot);
-	edit_status();			// show current file status
-	show_status_line();
 }
 
 static void dot_scroll(int cnt, int dir)
@@ -1416,8 +1402,6 @@
 	if (dot > q)
 		dot = begin_line(q);	// is dot is below bottom line?
 	dot_skip_over_ws();
-	edit_status();			// show current file status
-	show_status_line();
 }
 
 static void dot_skip_over_ws(void)
@@ -1608,11 +1592,11 @@
 		cmd_mode = 0;
 		cmdcnt = 0;
 		end_cmd_q();	// stop adding to q
-		*status_buffer = '\0';  // clear the status buffer
+		strcpy((char *) status_buffer, " ");	// clear the status buffer
 		if ((p[-1] != '\n') && (dot>text)) {
 			p--;
 		}
-	} else if (c == erase_char || c == 8 || c == 127) { // Is this a BS
+	} else if (c == erase_char) {	// Is this a BS
 		//     123456789
 		if ((p[-1] != '\n') && (dot>text)) {
 			p--;
@@ -2356,7 +2340,7 @@
 }
 
 //----- IO Routines --------------------------------------------
-static Byte get_one_char(void)
+static Byte get_one_char()
 {
 	static Byte c;
 
@@ -2616,25 +2600,25 @@
 }
 
 //----- Erase from cursor to end of line -----------------------
-static void clear_to_eol(void)
+static void clear_to_eol()
 {
 	write1(Ceol);   // Erase from cursor to end of line
 }
 
 //----- Erase from cursor to end of screen -----------------------
-static void clear_to_eos(void)
+static void clear_to_eos()
 {
 	write1(Ceos);   // Erase from cursor to end of screen
 }
 
 //----- Start standout mode ------------------------------------
-static void standout_start(void) // send "start reverse video" sequence
+static void standout_start() // send "start reverse video" sequence
 {
 	write1(SOs);     // Start reverse video mode
 }
 
 //----- End standout mode --------------------------------------
-static void standout_end(void) // send "end reverse video" sequence
+static void standout_end() // send "end reverse video" sequence
 {
 	write1(SOn);     // End reverse video mode
 }
@@ -2664,7 +2648,7 @@
 
 //----- Screen[] Routines --------------------------------------
 //----- Erase the Screen[] memory ------------------------------
-static void screen_erase(void)
+static void screen_erase()
 {
 	memset(screen, ' ', screensize);	// clear new screen
 }
@@ -2675,7 +2659,7 @@
 	static int last_cksum;
 	int l, cnt, cksum;
 
-	//edit_status();
+	edit_status();
 	cnt = strlen((char *) status_buffer);
 	for (cksum= l= 0; l < cnt; l++) { cksum += (int)(status_buffer[l]); }
 	// don't write the status line unless it changes
@@ -2697,10 +2681,10 @@
 
 	va_start(args, format);
 	strcpy((char *) status_buffer, SOs);	// Terminal standout mode on
-	vsprintf((char *) status_buffer + strlen((char *) status_buffer), format, args);
+	vsprintf((char *) status_buffer + strlen((char *) status_buffer), format,
+			 args);
 	strcat((char *) status_buffer, SOn);	// Terminal standout mode off
 	va_end(args);
-	show_status_line();
 
 	return;
 }
@@ -2713,7 +2697,6 @@
 	va_start(args, format);
 	vsprintf((char *) status_buffer, format, args);
 	va_end(args);
-	show_status_line();
 	return;
 }
 
@@ -2740,19 +2723,17 @@
 		cur = tot = 0;
 		percent = 100;
 	}
-
-	sprintf((char *) status_buffer,
-			"\"%s\""
+	psb("\"%s\""
 #ifdef CONFIG_FEATURE_VI_READONLY
-			"%s"
+		"%s"
 #endif							/* CONFIG_FEATURE_VI_READONLY */
-			"%s line %d of %d --%d%%--",
-			(cfn != 0 ? (char *) cfn : "No file"),
+		"%s line %d of %d --%d%%--",
+		(cfn != 0 ? (char *) cfn : "No file"),
 #ifdef CONFIG_FEATURE_VI_READONLY
-			((vi_readonly || readonly) ? " [Read only]" : ""),
+		((vi_readonly || readonly) ? " [Read only]" : ""),
 #endif							/* CONFIG_FEATURE_VI_READONLY */
-			(file_modified ? " [modified]" : ""),
-			cur, tot, percent);
+		(file_modified ? " [modified]" : ""),
+		cur, tot, percent);
 }
 
 //----- Force refresh of all Lines -----------------------------
@@ -2898,8 +2879,6 @@
 #endif							/* CONFIG_FEATURE_VI_OPTIMIZE_CURSOR */
 			}
 
-			edit_status();			// show current file status
-			show_status_line();
 			// write line out to terminal
 			{
 				int nic = ce-cs+1;
@@ -3079,18 +3058,15 @@
 		break;
 	case 7:			// ctrl-G  show current status
 		edit_status();
-		show_status_line();
 		break;
 	case 'h':			// h- move left
 	case VI_K_LEFT:	// cursor key Left
-	case 8:		// ctrl-H- move left    (This may be ERASE char)
-	case 127:	// DEL- move left   (This may be ERASE char)
+	case 8:			// ctrl-H- move left    (This may be ERASE char)
+	case 127:			// DEL- move left   (This may be ERASE char)
 		if (cmdcnt-- > 1) {
 			do_cmd(c);
 		}				// repeat cnt
 		dot_left();
-		edit_status();			// show current file status
-		show_status_line();
 		break;
 	case 10:			// Newline ^J
 	case 'j':			// j- goto next line, same col
@@ -3249,7 +3225,7 @@
 		//
 		// dont separate these two commands. 'f' depends on ';'
 		//
-		//**** fall thru to ... ';'
+		//**** fall thru to ... 'i'
 	case ';':			// ;- look at rest of line for last forward char
 		if (cmdcnt-- > 1) {
 			do_cmd(';');
@@ -3407,7 +3383,6 @@
 			}
 		} else if (strncasecmp((char *) p, "file", cnt) == 0 ) {
 			edit_status();			// show current file status
-			show_status_line();
 		} else if (sscanf((char *) p, "%d", &j) > 0) {
 			dot = find_line(j);		// go to line # j
 			dot_skip_over_ws();
@@ -3517,7 +3492,6 @@
 		dot_end();		// move to NL
 		if (dot < end - 1) {	// make sure not last char in text[]
 			*dot++ = ' ';	// replace NL with space
-			file_modified = TRUE;
 			while (isblnk(*dot)) {	// delete leading WS
 				dot_delete();
 			}
diff -Naur busybox-1.01.orig/include/usage.h busybox-1.01/include/usage.h
--- busybox-1.01.orig/include/usage.h	2005-08-17 09:29:15.000000000 +0800
+++ busybox-1.01/include/usage.h	2009-12-31 08:58:48.760428196 +0800
@@ -1886,8 +1886,15 @@
 	"Options:\n" \
 	"\t-c COUNT\tSend only COUNT pings.\n" \
 	"\t-s SIZE\t\tSend SIZE data bytes in packets (default=56).\n" \
-	"\t-q\t\tQuiet mode, only displays output at start\n" \
-	"\t\t\tand when finished."
+	"\t-q\t\tQuiet mode, only displays output at start and when finished.\n"\
+	"\t-W TIMEOUT\tTime to wait for a response, in seconds. The option affects only\n"\
+    "\t          \ttimeout in absense of any responses, otherwise  ping  waits  for\n"\
+    "\t          \ttwo RTTs."\
+	"\t-w DEADLINE\tSpecify  a  timeout, in seconds, before ping exits regardless of\n"\
+    "\t           \thow many packets have been sent or received. In this  case  ping\n"\
+    "\t           \tdoes  not  stop after count packet are sent, it waits either for\n"\
+    "\t           \tdeadline expire or until count probes are answered or  for  some\n"\
+    "\t           \terror notification from network.\n"
 #endif
 #define ping_example_usage \
 	"$ ping localhost\n" \
@@ -1978,12 +1985,13 @@
 	"/root\n"
 
 #define rdate_trivial_usage \
-	"[-sp] HOST"
+	"[-s] [-p] [-T <timeout>] HOST"
 #define rdate_full_usage \
 	"Get and possibly set the system date and time from a remote HOST.\n\n" \
 	"Options:\n" \
 	"\t-s\tSet the system date and time (default).\n" \
-	"\t-p\tPrint the date and time."
+	"\t-p\tPrint the date and time.\n" \
+	"\t-T TIMEOUT\tTime to wait for a response, in seconds. default is 10."
 
 #define readlink_trivial_usage \
 	""
@@ -2814,7 +2822,8 @@
 
 #define wget_trivial_usage \
 	"[-c|--continue] [-q|--quiet] [-O|--output-document file]\n" \
-	"\t\t[--header 'header: value'] [-Y|--proxy on/off] [-P DIR] url"
+	"\t\t[--header 'header: value'] [-Y|--proxy on/off] [-P DIR]\n" \
+	"\t\t[-U|--user-agent agent] [-T timeout] url"
 #define wget_full_usage \
 	"wget retrieves files via HTTP or FTP\n\n" \
 	"Options:\n" \
@@ -2822,6 +2831,8 @@
 	"\t-q\tquiet mode - do not print\n" \
 	"\t-P\tSet directory prefix to DIR\n" \
 	"\t-O\tsave to filename ('-' for stdout)\n" \
+	"\t-T\ttimeout,in seconds\n" \
+	"\t-U\tadjust 'User-Agent' field\n" \
 	"\t-Y\tuse proxy ('on' or 'off')"
 
 #define which_trivial_usage \
diff -Naur busybox-1.01.orig/init/init.c busybox-1.01/init/init.c
--- busybox-1.01.orig/init/init.c	2005-08-17 09:29:16.000000000 +0800
+++ busybox-1.01/init/init.c	2009-12-31 08:58:48.764430382 +0800
@@ -375,7 +375,7 @@
 	while ((fd = open(console, O_RDONLY | O_NONBLOCK)) < 0 && tried < 2) {
 		/* Can't open selected console -- try
 			logical system console and VT_MASTER */
-		safe_strncpy(console, (tried == 0 ? _PATH_CONSOLE : CURRENT_VC),
+		safe_strncpy(console, (tried == 0 ? _PATH_CONSOLE : "/dev/tty0"),
 							sizeof(console));
 		tried++;
 	}
@@ -632,15 +632,12 @@
 
 	pid = run(a);
 	while (1) {
-		wpid = waitpid(pid,&status,0);
+		wpid = wait(&status);
+		if (wpid > 0 && wpid != pid) {
+			continue;
+		}
 		if (wpid == pid)
 			break;
-		if (wpid == -1 && errno == ECHILD) {
-			/* we missed its termination */
-			break;
-		}
-		/* FIXME other errors should maybe trigger an error, but allow
-		 * the program to continue */
 	}
 	return wpid;
 }
diff -Naur busybox-1.01.orig/libbb/messages.c busybox-1.01/libbb/messages.c
--- busybox-1.01.orig/libbb/messages.c	2005-08-17 09:29:14.000000000 +0800
+++ busybox-1.01/libbb/messages.c	2009-12-31 08:58:48.764430382 +0800
@@ -59,27 +59,47 @@
 #endif
 
 #ifdef L_passwd_file
-#define PASSWD_FILE        "/etc/passwd"
+#ifdef CONFIG_AUTH_IN_VAR_ETC
+ #define PASSWD_FILE        "/var/etc/passwd"
+#else
+ #define PASSWD_FILE        "/etc/passwd"
+#endif
 const char * const bb_path_passwd_file = PASSWD_FILE;
 #endif
 
 #ifdef L_shadow_file
-#define SHADOW_FILE        "/etc/shadow"
+#ifdef CONFIG_AUTH_IN_VAR_ETC
+ #define SHADOW_FILE        "/var/etc/shadow"
+#else
+ #define SHADOW_FILE        "/etc/shadow"
+#endif
 const char * const bb_path_shadow_file = SHADOW_FILE;
 #endif
 
 #ifdef L_group_file
-#define GROUP_FILE         "/etc/group"
+#ifdef CONFIG_AUTH_IN_VAR_ETC
+ #define GROUP_FILE         "/var/etc/group"
+#else
+ #define GROUP_FILE         "/etc/group"
+#endif
 const char * const bb_path_group_file = GROUP_FILE;
 #endif
 
 #ifdef L_gshadow_file
-#define GSHADOW_FILE       "/etc/gshadow"
+#ifdef CONFIG_AUTH_IN_VAR_ETC
+ #define GSHADOW_FILE       "/var/etc/gshadow"
+#else
+ #define GSHADOW_FILE       "/etc/gshadow"
+#endif
 const char * const bb_path_gshadow_file = GSHADOW_FILE;
 #endif
 
 #ifdef L_nologin_file
-#define NOLOGIN_FILE       "/etc/nologin"
+#ifdef CONFIG_AUTH_IN_VAR_ETC
+ #define NOLOGIN_FILE       "/var/etc/nologin"
+#else
+ #define NOLOGIN_FILE       "/etc/nologin"
+#endif
 const char * const bb_path_nologin_file = NOLOGIN_FILE;
 #endif
 
diff -Naur busybox-1.01.orig/libbb/procps.c busybox-1.01/libbb/procps.c
--- busybox-1.01.orig/libbb/procps.c	2005-08-17 09:29:14.000000000 +0800
+++ busybox-1.01/libbb/procps.c	2009-12-31 08:58:48.764430382 +0800
@@ -22,7 +22,7 @@
 #endif
 	)
 {
-	static DIR *dir;
+	static DIR *dir, *taskdir;
 	struct dirent *entry;
 	static procps_status_t ret_status;
 	char *name;
@@ -32,6 +32,7 @@
 	FILE *fp;
 	procps_status_t curstatus;
 	int pid;
+	static int parentpid;
 	long tasknice;
 	struct stat sb;
 
@@ -40,111 +41,150 @@
 		if(!dir)
 			bb_error_msg_and_die("Can't open /proc");
 	}
-	for(;;) {
-		if((entry = readdir(dir)) == NULL) {
-			closedir(dir);
-			dir = 0;
-			return 0;
-		}
-		name = entry->d_name;
-		if (!(*name >= '0' && *name <= '9'))
-			continue;
 
-		memset(&curstatus, 0, sizeof(procps_status_t));
-		pid = atoi(name);
-		curstatus.pid = pid;
+	if (taskdir) {
+		for(;;) {
+			if((entry = readdir(taskdir)) == NULL) {
+				closedir(taskdir);
+				taskdir = 0;
+				break;
+			}
+			name = entry->d_name;
+			if (!(*name >= '0' && *name <= '9'))
+				continue;
 
-		sprintf(status, "/proc/%d", pid);
-		if(stat(status, &sb))
-			continue;
-		my_getpwuid(curstatus.user, sb.st_uid, sizeof(curstatus.user));
+			memset(&curstatus, 0, sizeof(procps_status_t));
+			pid = atoi(name);
+			curstatus.pid = pid;
 
-		sprintf(status, "/proc/%d/stat", pid);
-		if((fp = fopen(status, "r")) == NULL)
-			continue;
-#ifdef CONFIG_SELINUX
-		if(use_selinux)
-		{
-			if(fstat_secure(fileno(fp), &sb, sid))
+			sprintf(status, "/proc/%d/task/%d", parentpid, pid);
+			if(stat(status, &sb))
 				continue;
-		}
-		else
+			my_getpwuid(curstatus.user, sb.st_uid, sizeof(curstatus.user));
+
+			sprintf(status, "/proc/%d/task/%d/stat", parentpid, pid);
+			if((fp = fopen(status, "r")) == NULL)
+				continue;
+#ifdef CONFIG_SELINUX
+			if(use_selinux)
+			{
+				if(fstat_secure(fileno(fp), &sb, sid))
+					continue;
+			}
+			else
 #endif
-		name = fgets(buf, sizeof(buf), fp);
-		fclose(fp);
-		if(name == NULL)
-			continue;
-		name = strrchr(buf, ')'); /* split into "PID (cmd" and "<rest>" */
-		if(name == 0 || name[1] != ' ')
-			continue;
-		*name = 0;
-		sscanf(buf, "%*s (%15c", curstatus.short_cmd);
-		n = sscanf(name+2,
-		"%c %d "
-		"%*s %*s %*s %*s "     /* pgrp, session, tty, tpgid */
-		"%*s %*s %*s %*s %*s " /* flags, min_flt, cmin_flt, maj_flt, cmaj_flt */
+			name = fgets(buf, sizeof(buf), fp);
+			fclose(fp);
+			if(name == NULL)
+				continue;
+			name = strrchr(buf, ')'); /* split into "PID (cmd" and "<rest>" */
+			if(name == 0 || name[1] != ' ')
+				continue;
+			*name = 0;
+			sscanf(buf, "%*s (%15c", curstatus.short_cmd);
+			n = sscanf(name+2,
+			"%c %d "
+			"%*s %*s %*s %*s "     /* pgrp, session, tty, tpgid */
+			"%*s %*s %*s %*s %*s " /* flags, min_flt, cmin_flt, maj_flt, cmaj_flt */
 #ifdef FEATURE_CPU_USAGE_PERCENTAGE
-		"%lu %lu "
+			"%lu %lu "
 #else
-		"%*s %*s "
+			"%*s %*s "
 #endif
-		"%*s %*s %*s "         /* cutime, cstime, priority */
-		"%ld "
-		"%*s %*s %*s "         /* timeout, it_real_value, start_time */
-		"%*s "                 /* vsize */
-		"%ld",
-		curstatus.state, &curstatus.ppid,
+			"%*s %*s %*s "         /* cutime, cstime, priority */
+			"%ld "
+			"%*s %*s %*s "         /* timeout, it_real_value, start_time */
+			"%*s "                 /* vsize */
+			"%ld",
+			curstatus.state, &curstatus.ppid,
 #ifdef FEATURE_CPU_USAGE_PERCENTAGE
-		&curstatus.utime, &curstatus.stime,
+			&curstatus.utime, &curstatus.stime,
 #endif
-		&tasknice,
-		&curstatus.rss);
+			&tasknice,
+			&curstatus.rss);
+			if ( pid != parentpid )
+				curstatus.ppid = parentpid;
 #ifdef FEATURE_CPU_USAGE_PERCENTAGE
-		if(n != 6)
+			if(n != 6)
 #else
-		if(n != 4)
+			if(n != 4)
 #endif
-			continue;
+				continue;
 
-		if (curstatus.rss == 0 && curstatus.state[0] != 'Z')
-			curstatus.state[1] = 'W';
-		else
-			curstatus.state[1] = ' ';
-		if (tasknice < 0)
-			curstatus.state[2] = '<';
-		else if (tasknice > 0)
-			curstatus.state[2] = 'N';
-		else
-			curstatus.state[2] = ' ';
+			if (curstatus.rss == 0 && curstatus.state[0] != 'Z')
+				curstatus.state[1] = 'W';
+			else
+				curstatus.state[1] = ' ';
+			if (tasknice < 0)
+				curstatus.state[2] = '<';
+			else if (tasknice > 0)
+				curstatus.state[2] = 'N';
+			else
+				curstatus.state[2] = ' ';
 
 #ifdef PAGE_SHIFT
-		curstatus.rss <<= (PAGE_SHIFT - 10);     /* 2**10 = 1kb */
+			curstatus.rss <<= (PAGE_SHIFT - 10);     /* 2**10 = 1kb */
 #else
-		curstatus.rss *= (getpagesize() >> 10);     /* 2**10 = 1kb */
+			curstatus.rss *= (getpagesize() >> 10);     /* 2**10 = 1kb */
 #endif
 
-		if(save_user_arg0) {
-			sprintf(status, "/proc/%d/cmdline", pid);
-			if((fp = fopen(status, "r")) == NULL)
-				continue;
-			if((n=fread(buf, 1, sizeof(buf)-1, fp)) > 0) {
-				if(buf[n-1]=='\n')
-					buf[--n] = 0;
-				name = buf;
-				while(n) {
-					if(((unsigned char)*name) < ' ')
-						*name = ' ';
-					name++;
-					n--;
+			if(save_user_arg0) {
+				sprintf(status, "/proc/%d/task/%d/cmdline", parentpid, pid);
+				if((fp = fopen(status, "r")) == NULL)
+					continue;
+				if((n=fread(buf, 1, sizeof(buf)-1, fp)) > 0) {
+					if(buf[n-1]=='\n')
+						buf[--n] = 0;
+					name = buf;
+					while(n) {
+						if(((unsigned char)*name) < ' ')
+							*name = ' ';
+						name++;
+						n--;
+					}
+					*name = 0;
+					if(buf[0])
+						curstatus.cmd = strdup(buf);
+					/* if NULL it work true also */
 				}
-				*name = 0;
-				if(buf[0])
-					curstatus.cmd = strdup(buf);
-				/* if NULL it work true also */
+				fclose(fp);
 			}
-			fclose(fp);
+			return memcpy(&ret_status, &curstatus, sizeof(procps_status_t));
+		}
+	}
+
+	for(;;) {
+		if((entry = readdir(dir)) == NULL) {
+			closedir(dir);
+			dir = 0;
+			return 0;
+		}
+		name = entry->d_name;
+		if (!(*name >= '0' && *name <= '9'))
+			continue;
+
+		memset(&curstatus, 0, sizeof(procps_status_t));
+		pid = atoi(name);
+		curstatus.pid = pid;
+
+		sprintf(status, "/proc/%d", pid);
+		if(stat(status, &sb))
+			continue;
+		my_getpwuid(curstatus.user, sb.st_uid, sizeof(curstatus.user));
+
+		if (!taskdir) {
+			sprintf(status, "/proc/%d/task", pid);
+			taskdir = opendir(status);
+			if(!taskdir)
+					// ignore processes without taskdir
+				continue;
+			parentpid = pid;
+			return procps_scan(save_user_arg0
+#ifdef CONFIG_SELINUX
+						,use_selinux, sid
+#endif
+				);
 		}
-		return memcpy(&ret_status, &curstatus, sizeof(procps_status_t));
 	}
 }
 
diff -Naur busybox-1.01.orig/miscutils/hdparm.c busybox-1.01/miscutils/hdparm.c
--- busybox-1.01.orig/miscutils/hdparm.c	2005-08-17 09:29:15.000000000 +0800
+++ busybox-1.01/miscutils/hdparm.c	2009-12-31 08:58:48.768429495 +0800
@@ -1254,6 +1254,7 @@
 static unsigned long set_sleepnow   = 0, get_sleepnow   = 0;
 static unsigned long get_powermode  = 0;
 static unsigned long set_apmmode = 0, get_apmmode= 0, apmmode = 0;
+static unsigned long set_acoustic = 0, get_acoustic = 0, acoustic = 0;
 #endif
 #ifdef CONFIG_FEATURE_HDPARM_GET_IDENTITY
 static int get_IDentity = 0;
@@ -2123,6 +2124,20 @@
 		}
 		bb_ioctl(fd, HDIO_DRIVE_CMD, &args,"HDIO_DRIVE_CMD");
 	}
+	if (set_acoustic)
+	{
+		no_scsi();
+		acoustic=check_if_min_and_set_val(acoustic,0);
+		acoustic=check_if_maj_and_set_val(acoustic,254);
+		if_printf(get_acoustic," setting AAM level to 0x%02lX (%ld)\n", acoustic, acoustic);
+		bb_ioctl(fd, HDIO_SET_ACOUSTIC, (int*)acoustic,"HDIO_SET_ACOUSTIC");
+	}
+	if (get_acoustic)
+	{
+		no_scsi();
+		bb_ioctl(fd, HDIO_GET_ACOUSTIC, (unsigned long*)&parm,"HDIO_GET_ACOUSTIC");
+		printf(" acoustic     = %2ld (128=quiet ... 254=fast)\n", parm);
+	}
 	if (set_wcache)
 	{
 #ifdef DO_FLUSHCACHE
@@ -2831,6 +2846,13 @@
 							p = *argv++, --argc;
 						p=GET_NUMBER(p,&set_readahead,&Xreadahead);
 						break;
+					case 'M':
+						get_acoustic = noisy; 
+						noisy = 1;
+						if (!*p && argc && isalnum(**argv))
+							p = *argv++, --argc;
+						p=GET_NUMBER(p,&set_acoustic,&acoustic);
+						break;
 					case 'B':
 						get_apmmode = noisy;
 						noisy = 1;
diff -Naur busybox-1.01.orig/modutils/lsmod.c busybox-1.01/modutils/lsmod.c
--- busybox-1.01.orig/modutils/lsmod.c	2005-08-17 09:29:11.000000000 +0800
+++ busybox-1.01/modutils/lsmod.c	2009-12-31 08:58:48.768429495 +0800
@@ -38,6 +38,9 @@
 #include <sys/file.h>
 #include "busybox.h"
 
+#ifdef CONFIG_FEATURE_2_6_MODULES
+#undef CONFIG_FEATURE_QUERY_MODULE_INTERFACE
+#endif
 
 #ifndef CONFIG_FEATURE_CHECK_TAINTED_MODULE
 static inline void check_tainted(void) { printf("\n"); }
diff -Naur busybox-1.01.orig/networking/ping.c busybox-1.01/networking/ping.c
--- busybox-1.01.orig/networking/ping.c	2005-08-17 09:29:10.000000000 +0800
+++ busybox-1.01/networking/ping.c	2009-12-31 08:58:48.772429726 +0800
@@ -57,8 +57,8 @@
 static const int MAXICMPLEN = 76;
 static const int MAXPACKET = 65468;
 #define	MAX_DUP_CHK	(8 * 128)
-static const int MAXWAIT = 10;
-static const int PINGINTERVAL = 1;		/* second */
+static const long MAXWAIT = 10;
+static const long PINGINTERVAL = 1;		/* 1/1000 second */
 
 #define O_QUIET         (1 << 0)
 
@@ -71,6 +71,7 @@
 static void ping(const char *host);
 
 /* common routines */
+
 static int in_cksum(unsigned short *buf, int sz)
 {
 	int nleft = sz;
@@ -133,6 +134,7 @@
 
 	signal(SIGALRM, noresp);
 	alarm(5);					/* give the host 5000ms to respond */
+
 	/* listen for replies */
 	while (1) {
 		struct sockaddr_in from;
@@ -174,6 +176,7 @@
 static int datalen; /* intentionally uninitialized to work around gcc bug */
 
 static long ntransmitted, nreceived, nrepeats, pingcount;
+static unsigned total_secs,timeout,deadline;
 static int myid, options;
 static unsigned long tmin = ULONG_MAX, tmax, tsum;
 static char rcvd_tbl[MAX_DUP_CHK / 8];
@@ -239,13 +242,29 @@
 		bb_error_msg_and_die("ping wrote %d chars; %d expected", i,
 			   (int)sizeof(packet));
 
-	signal(SIGALRM, sendping);
-	if (pingcount == 0 || ntransmitted < pingcount) {	/* schedule next in 1s */
+	if (pingcount == 0 || deadline || ntransmitted < pingcount) {	/* schedule next in 1s */
+		signal(SIGALRM, sendping);
+		if (deadline) {
+			total_secs += PINGINTERVAL;
+			if (total_secs >= deadline)
+				signal(SIGALRM, pingstats);
+		}
 		alarm(PINGINTERVAL);
-	} else {					/* done, wait for the last ping to come back */
-		/* todo, don't necessarily need to wait so long... */
+	} else {/* -c NN, and all NN are sent (and no deadline) */
+		/* Wait for the last ping to come back.
+		 * -W timeout: wait for a response in seconds.
+		 * Affects only timeout in absense of any responses,
+		 * otherwise ping waits for two RTTs. */
+		unsigned expire = timeout;
+
+		if (nreceived) {
+			/* approx. 2*tmax, in seconds (2 RTT) */
+			expire = tmax / (512*1024);
+			if (expire == 0)
+				expire = 1;
+		}
 		signal(SIGALRM, pingstats);
-		alarm(MAXWAIT);
+		alarm(expire);
 	}
 }
 
@@ -400,6 +419,8 @@
 	argc--;
 	argv++;
 	options = 0;
+	timeout = MAXWAIT;
+	deadline=0;
 	/* Parse any options */
 	while (argc >= 1 && **argv == '-') {
 		thisarg = *argv;
@@ -420,6 +441,18 @@
 			argv++;
 			datalen = atoi(*argv);
 			break;
+		case 'w':
+			if (--argc <=0)
+					bb_show_usage();
+			argv++;
+			deadline= atoi(*argv);
+			break;
+		case 'W':
+			if (--argc <=0)
+					bb_show_usage();
+			argv++;
+			timeout= atoi(*argv);
+			break;
 		default:
 			bb_show_usage();
 		}
diff -Naur busybox-1.01.orig/networking/wget.c busybox-1.01/networking/wget.c
--- busybox-1.01.orig/networking/wget.c	2005-08-17 09:29:10.000000000 +0800
+++ busybox-1.01/networking/wget.c	2009-12-31 09:02:32.668431592 +0800
@@ -40,6 +40,7 @@
 };
 
 static void parse_url(char *url, struct host_info *h);
+static void exit_timeout(int ign);
 static FILE *open_socket(struct sockaddr_in *s_in);
 static char *gethdr(char *buf, size_t bufsiz, FILE *fp, int *istrunc);
 static int ftpcmd(char *s1, char *s2, FILE *fp, char *buf);
@@ -51,6 +52,7 @@
 static void progressmeter(int flag);
 static char *curfile;			/* Name of current file being transferred. */
 static struct timeval start;	/* Time a transfer started. */
+static int timeout;
 static volatile unsigned long statbytes = 0; /* Number of bytes transferred so far. */
 /* For progressmeter() -- number of seconds before xfer considered "stalled" */
 static const int STALLTIME = 5;
@@ -149,6 +151,8 @@
 #define WGET_OPT_HEADER	16
 #define WGET_OPT_PREFIX	32
 #define WGET_OPT_PROXY	64
+#define WGET_OPT_USER_AGENT    128
+#define WGET_OPT_NETWORK_READ_TIMEOUT 256
 
 static const struct option wget_long_options[] = {
 	{ "continue",        0, NULL, 'c' },
@@ -158,6 +162,8 @@
 	{ "header",	         1, NULL, 131 },
 	{ "directory-prefix",1, NULL, 'P' },
 	{ "proxy",           1, NULL, 'Y' },
+	{ "user-agent",      1, NULL, 'U' },
+	{ "timeout",		 1, NULL, 'T' },
 	{ 0,                 0, 0, 0 }
 };
 
@@ -187,13 +193,16 @@
 	int quiet_flag = FALSE;		/* Be verry, verry quiet...			*/
 	int use_proxy = 1;          /* Use proxies if env vars are set  */
 	char *proxy_flag = "on";	/* Use proxies if env vars are set  */
+	char *user_agent = "Wget"; /* Content of the "User-Agent" header field */
+	char * ctimeout=0;
+	timeout=0;
 
 	/*
 	 * Crack command line.
 	 */
 	bb_opt_complementaly = "\203*";
 	bb_applet_long_options = wget_long_options;
-	opt = bb_getopt_ulflags(argc, argv, "cq\213O:\203:P:Y:", &fname_out, &headers_llist, &dir_prefix, &proxy_flag);
+	opt = bb_getopt_ulflags(argc, argv, "cq\213O:\203:P:Y:U:T:", &fname_out, &headers_llist, &dir_prefix, &proxy_flag, &user_agent,&ctimeout);
 	if (opt & WGET_OPT_CONTINUE) {
 		++do_continue;
 	}
@@ -218,6 +227,14 @@
 			headers_llist = headers_llist->link;
 		}
 	}
+	if(opt & WGET_OPT_NETWORK_READ_TIMEOUT){
+		timeout=atoi(ctimeout);
+		if(timeout>0){
+			signal(SIGALRM, exit_timeout);
+			alarm(timeout);					
+
+		}
+	}
 	if (argc - optind != 1)
 			bb_show_usage();
 
@@ -330,7 +347,7 @@
 				fprintf(sfp, "GET /%s HTTP/1.1\r\n", target.path);
 			}
 
-			fprintf(sfp, "Host: %s\r\nUser-Agent: Wget\r\n", target.host);
+			fprintf(sfp, "Host: %s\r\nUser-Agent: %s\r\n", target.host, user_agent);
 
 #ifdef CONFIG_FEATURE_WGET_AUTHENTICATION
 			if (target.user) {
@@ -602,6 +619,11 @@
 	}
 }
 
+void exit_timeout(int ign)
+{
+	printf("timeout,exit...");
+	exit(EXIT_FAILURE);
+}
 
 FILE *open_socket(struct sockaddr_in *s_in)
 {
diff -Naur busybox-1.01.orig/sysdeps/linux/Config.in busybox-1.01/sysdeps/linux/Config.in
--- busybox-1.01.orig/sysdeps/linux/Config.in	2005-08-17 09:29:15.000000000 +0800
+++ busybox-1.01/sysdeps/linux/Config.in	2009-12-31 08:58:48.772429726 +0800
@@ -203,6 +203,14 @@
 	  if you want to add some simple compiler switches (like -march=i686),
 	  or check for warnings using -Werror, just those options here.
 
+config CONFIG_AUTH_IN_VAR_ETC
+	bool "Search auth files in /var/etc"
+	default n
+	help
+	  passwd, shadow, group, gshadow and nologin will be searched in
+	  /var/etc instead of /etc. This is usefull, if /etc is not
+	  writeable (flash).
+
 endmenu
 
 menu 'Installation Options'
diff -Naur busybox-1.01.orig/util-linux/mount.c busybox-1.01/util-linux/mount.c
--- busybox-1.01.orig/util-linux/mount.c	2005-08-17 09:29:17.000000000 +0800
+++ busybox-1.01/util-linux/mount.c	2009-12-31 08:58:48.776424648 +0800
@@ -52,6 +52,11 @@
 #include <stdio.h>
 #include <mntent.h>
 #include <ctype.h>
+#include <netdb.h>
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <arpa/inet.h>
+
 #include "busybox.h"
 
 #ifdef CONFIG_NFSMOUNT
@@ -475,6 +480,30 @@
 				}
 			}
 #endif
+			if (device[0] == '/' && device[1] == '/') {
+				/* CIFS or SMBFS mount. Do hostname resolving ourselves */
+				struct hostent *s_addr;
+				char *sl = strchr(device+2, '/');
+				char *unc = strstr(string_flags, "unc=//");
+				char *slunc;
+				char *smbdev = xmalloc(PATH_MAX);
+				char *newopts = xmalloc(PATH_MAX);
+
+				strncpy(smbdev, device+2, sl-(device+2));
+				s_addr = gethostbyname(smbdev);
+				
+				snprintf(smbdev, PATH_MAX, "//%s%s", inet_ntoa(*( struct in_addr*)( s_addr->h_addr)), sl);
+				free(device);
+				device = smbdev;
+				
+				if (unc) {
+					slunc = strchr(unc+6, '/');
+					string_flags[unc - string_flags] = '\0';
+					snprintf(newopts, PATH_MAX, "%sunc=//%s%s", string_flags, inet_ntoa(*( struct in_addr*)( s_addr->h_addr)), slunc);
+					free(string_flags);
+					string_flags = newopts;
+				}
+			}
 			if (!mount_one
 				(device, directory, filesystemType, flags, string_flags,
 				 useMtab, fakeIt, extra_opts, TRUE, all)) {
diff -Naur busybox-1.01.orig/util-linux/rdate.c busybox-1.01/util-linux/rdate.c
--- busybox-1.01.orig/util-linux/rdate.c	2005-08-17 09:29:17.000000000 +0800
+++ busybox-1.01/util-linux/rdate.c	2009-12-31 08:58:48.776424648 +0800
@@ -38,6 +38,7 @@
 
 
 static const int RFC_868_BIAS = 2208988800UL;
+static int timeout;
 
 static void socket_timeout(int sig)
 {
@@ -54,7 +55,7 @@
 	s_in.sin_port = bb_lookup_port("time", "tcp", 37);
 
 	/* Add a timeout for dead or non accessable servers */
-	alarm(10);
+	alarm(timeout);
 	signal(SIGALRM, socket_timeout);
 
 	fd = xconnect(&s_in);
@@ -77,31 +78,50 @@
 
 int rdate_main(int argc, char **argv)
 {
+	char *thisarg;
+
+	argc--;
+	argv++;
+
 	time_t remote_time;
 	int opt;
 	int setdate = 1;
 	int printdate = 1;
-
-	/* Interpret command line args */
-	while ((opt = getopt(argc, argv, "sp")) > 0) {
-		switch (opt) {
+	timeout=10;
+	int opt_s=0;
+	int opt_p=0;
+
+	while (argc >= 1 && **argv == '-') {
+		thisarg = *argv;
+		thisarg++;
+		switch (*thisarg) {
 			case 's':
 				printdate = 0;
 				setdate = 1;
+				opt_s=1;
 				break;
 			case 'p':
 				printdate = 1;
 				setdate = 0;
+				opt_p=1;
+				break;
+			case 'T':
+				if (--argc <= 0)
+			        bb_show_usage();
+				argv++;
+				timeout=atoi(*argv);
 				break;
 			default:
 				bb_show_usage();
 		}
+		argc--;
+		argv++;
 	}
 
-	if (optind == argc)
+	if ((opt_s && opt_p) || argc <1 )
 		bb_show_usage();
 
-	remote_time = askremotedate(argv[optind]);
+	remote_time = askremotedate(*argv);
 
 	if (setdate) {
 		time_t current_time;
